### PR TITLE
start PR for etcd - enable authentication

### PR DESCRIPTION
**Pull request by anshupande**

### CODE REVIEW

**Comment by anshupande** [comment]
we will use something similar to get master/root credentials: https://github.com/adobe-community/issues/issues/8204

**Comment by matthewdfuller** [comment]
This is launching after various other services relying on etcd start, is that okay?

**Comment by anshupande** [comment]
since we have guest role already present we can run this anytime....Do we have any order that we should follow? Currently I have `After=docker.service fleet.service bootstrap.service
Requires=docker.service fleet.service bootstrap.service`

**Comment by anshupande** [comment]
@matthewdfuller Or we can add this service in every other service as after/required , just as we have docker.service required by every service?

File: v3/util/etcdctl-auth-root.sh [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+#Disable authentication
+
+#curl  -L http://127.0.0.1:2379/v2/auth/enable -XDELETE
+
+etcdctlrootusername=$(etcdctl get /etcdctl/config/root-user)
+etcdctlrootpassword=$(etcdctl get /etcdctl/config/root-password)
+etcdctletcdreadusername=$(etcdctl get /etcdctl/config/read-user)
+etcdctletcdreadpassword=$(etcdctl get /etcdctl/config/read-password)
+etcdctletcdreadwriteusername=$(etcdctl get /etcdctl/config/write-user)
+etcdctletcdreadwritepassword=$(etcdctl get /etcdctl/config/write-password)
```
Comment (matthewdfuller) [inline v3/util/etcdctl-auth-root.sh]: Nitpick, but can we replace these with uppercase equivalents and drop the etcdctl?

```
ROOT_USERNAME
ROOT_PASSWORD
READ_USERNAME
READ_PASSWORD
WRITE_USERNAME
WRITE_PASSWORD
```

Then update occurrences below.

File: v3/util/etcdctl-auth-root.sh [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+#add a root user
+
+echo '{"user":"'$(eval echo $etcdctlrootusername)'", "password":"'$(eval echo $etcdctlrootpassword)'"}' > /home/core/mesos-systemd/v3/util/root.json
```
Comment (matthewdfuller) [inline v3/util/etcdctl-auth-root.sh]: Change `/home/core/mesos-systemd/v3/util/root.json` to `/opt/etcdctl/root-user.json`

File: v3/util/etcdctl-auth-root.sh [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+curl  -L http://127.0.0.1:2379/v2/auth/users/$(eval echo $etcdctlrootusername) -XPUT -d "@/home/core/mesos-systemd/v3/util/root.json"
+
+
+#add a non root read only user
+echo '{"user":"'$(eval echo $etcdctletcdreadusername)'", "password":"'$(eval echo $etcdctletcdreadpassword)'"}' > /home/core/mesos-systemd/v3/util/etcdreaduser.json
```
Comment (matthewdfuller) [inline v3/util/etcdctl-auth-root.sh]: Change `/home/core/mesos-systemd/v3/util/etcdreaduser.json` to `/opt/etcdctl/read-user.json`

File: v3/util/etcdctl-auth-root.sh [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+curl   -L http://127.0.0.1:2379/v2/auth/users/$(eval echo $etcdctletcdreadusername) -XPUT -d "@/home/core/mesos-systemd/v3/util/etcdreaduser.json"
+
+#add a non root readwrite user
+echo '{"user":"'$(eval echo $etcdctletcdreadwriteusername)'", "password":"'$(eval echo $etcdctletcdreadwritepassword)'"}' > /home/core/mesos-systemd/v3/util/etcdwriteuser.json
```
Comment (matthewdfuller) [inline v3/util/etcdctl-auth-root.sh]: Change `/home/core/mesos-systemd/v3/util/etcdwriteuser.json` to `/opt/etcdctl/write-user.json`

File: v3/util/etcdctl-auth-root.sh [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+curl  -L http://127.0.0.1:2379/v2/auth/users/$(eval echo $etcdctletcdreadwriteusername) -XPUT -d "@/home/core/mesos-systemd/v3/util/etcdwriteuser.json"
+
+#Create a read only role:
+
+etcdctl role add readonlyrolename
```
Comment (matthewdfuller) [inline v3/util/etcdctl-auth-root.sh]: Is this actually called "readonlyrolename"? Or should there be an actual name here such as "readonlyrole"?

File: v3/util/etcdctl-auth-root.sh (hunk 1 above)
Comment (matthewdfuller) [inline v3/util/etcdctl-auth-root.sh]: @chr0n1x do you agree with the above or is lowercase better here?

File: v3/util/etcdctl-auth-root.sh (hunk 5 above)
Comment (anshupande) [inline v3/util/etcdctl-auth-root.sh]: readonlyrole will work as well

**Comment by anshupande** [comment]
@matthewdfuller This is cleaned to follow naming convention. Let me know if we need to call it during bootstrap or running it as service is good?

**Comment by anshupande** [comment]
Authentication can be enabled just by running as part of bootstrap and hence I will delete the service ```core@ip-10-74-131-135 sandbox-anshu-us-east-1c-control /opt $ curl  -L http://127.0.0.1:2379/v2/auth/enable -XGET
{"enabled":true}
core@ip-10-74-131-135 sandbox-anshu-us-east-1c-control /opt $ curl  -L http://127.0.0.1:2379/v2/auth/users -XGET
{"message":"Insufficient credentials"}core@ip-10-74-131-135 sandbox-anshu-us-east-1c-control /opt $ 

```
```

File: v3/setup/etcdctl-auth.sh [hunk 6]
```diff
@@ -0,0 +1,82 @@
+#!/bin/bash
+
+source /etc/environment
+
+#Disable authentication
+
+#curl  -L http://127.0.0.1:2379/v2/auth/enable -XDELETE
```
Comment (chr0n1x) [inline v3/setup/etcdctl-auth.sh]: remove this comment

File: v3/setup/etcdctl-auth.sh [hunk 7]
```diff
@@ -0,0 +1,82 @@
+#!/bin/bash
+
+source /etc/environment
+
+#Disable authentication
```
Comment (chr0n1x) [inline v3/setup/etcdctl-auth.sh]: better comments please; this doesn't say much

File: v3/setup/etcdctl-auth.sh [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+WRITE_PASSWORD=$(etcdctl get /etcdctl/config/write-password)
+
+CRED_DIR="/opt/etcdctl"
+if [[ ! -d $CRED_DIR ]]; then
+    sudo mkdir $CRED_DIR -p
+fi
+
+#add a root user
+
+sudo echo '{"user":"'$(eval echo $ROOT_USERNAME)'", "password":"'$(eval echo $ROOT_PASSWORD)'"}' > $CRED_DIR/root-user.json
```
Comment (chr0n1x) [inline v3/setup/etcdctl-auth.sh]: I don't think you need the `eval`. Actually `"'${ROOT_PASSWORD}'"` might just work.

File: v3/setup/etcdctl-auth.sh [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+curl  -L http://127.0.0.1:2379/v2/auth/users/$(eval echo $ROOT_USERNAME) -XPUT -d "@$CRED_DIR/root-user.json"
```
Comment (chr0n1x) [inline v3/setup/etcdctl-auth.sh]: same here

File: v3/setup/etcdctl-auth.sh [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+
+
+#add a non root read only user
+sudo echo '{"user":"'$(eval echo $READ_USERNAME)'", "password":"'$(eval echo $READ_PASSWORD)'"}' > $CRED_DIR/read-user.json
```
Comment (chr0n1x) [inline v3/setup/etcdctl-auth.sh]: same here

File: v3/setup/etcdctl-auth.sh [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+curl   -L http://127.0.0.1:2379/v2/auth/users/$(eval echo $READ_USERNAME) -XPUT -d "@$CRED_DIR/read-user.json"
```
Comment (chr0n1x) [inline v3/setup/etcdctl-auth.sh]: and here

File: v3/setup/etcdctl-auth.sh [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+
+#add a non root readwrite user
+sudo echo '{"user":"'$(eval echo $WRITE_USERNAME)'", "password":"'$(eval echo $WRITE_PASSWORD)'"}' > $CRED_DIR/write-user.json
```
Comment (chr0n1x) [inline v3/setup/etcdctl-auth.sh]: and here

File: v3/setup/etcdctl-auth.sh [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+curl  -L http://127.0.0.1:2379/v2/auth/users/$(eval echo $WRITE_USERNAME) -XPUT -d "@$CRED_DIR/write-user.json"
```
Comment (chr0n1x) [inline v3/setup/etcdctl-auth.sh]: here

File: v3/setup/etcdctl-auth.sh [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+
+#Create a read only role:
+
+etcdctl role add readonlyrole
```
Comment (chr0n1x) [inline v3/setup/etcdctl-auth.sh]: are there any restricted characters on role names? That is, can we just do something like

```
etcdctl role add read-only
```

2 issues with the current state:
1. mashing the words together isn't really readable 
2. `read-only-role` has the word `role` in it. It's already evident, on setting & ls'ing roles that this "thing" is already a role, so the word `role` itself does not have to be in the name

File: v3/setup/etcdctl-auth.sh [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+
+#Give read only access to this role:
+
+etcdctl role grant readonlyrole -path '/*' -read
+
+#Create a readwrite only role:
+
+etcdctl role add readwriterole
```
Comment (chr0n1x) [inline v3/setup/etcdctl-auth.sh]: same here

File: v3/setup/etcdctl-auth.sh [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+
+#Give readwrite only access to this role:
+
+etcdctl role grant readwriterole -path '/*' -readwrite
+
+#Give read only non root user read only role
+
+sudo echo '{"user": "'$(eval echo $READ_USERNAME)'", "grant": ["readonlyrole"]}' > $CRED_DIR/readonlyrole.json
```
Comment (chr0n1x) [inline v3/setup/etcdctl-auth.sh]: remove `eval`

File: v3/setup/etcdctl-auth.sh [hunk 17]
```diff
@@ -0,0 +1,82 @@
...
+
+curl  -L http://127.0.0.1:2379/v2/auth/users/$(eval echo $READ_USERNAME) -XPUT -d "@$CRED_DIR/readonlyrole.json"
```
Comment (chr0n1x) [inline v3/setup/etcdctl-auth.sh]: same here

File: v3/setup/etcdctl-auth.sh [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+
+#Give readwrite non root user readwrite role
+
+sudo echo '{"user": "'$(eval echo $WRITE_USERNAME)'", "grant": ["readwriterole"]}' > $CRED_DIR/readwriterole.json
```
Comment (chr0n1x) [inline v3/setup/etcdctl-auth.sh]: and here

File: v3/setup/etcdctl-auth.sh [hunk 19]
```diff
@@ -0,0 +1,82 @@
...
+
+curl  -L http://127.0.0.1:2379/v2/auth/users/$(eval echo $WRITE_USERNAME) -XPUT -d "@$CRED_DIR/readwriterole.json"
```
Comment (chr0n1x) [inline v3/setup/etcdctl-auth.sh]: and here

File: v3/setup/etcdctl-auth.sh [hunk 20]
```diff
@@ -0,0 +1,82 @@
...
+
+#revoke guest role read:write access
```
Comment (chr0n1x) [inline v3/setup/etcdctl-auth.sh]: spaces after `#`

File: v3/setup/etcdctl-auth.sh [hunk 21]
```diff
@@ -0,0 +1,82 @@
...
+
+#etcdctl role revoke guest -path '/*' -readwrite
```
Comment (chr0n1x) [inline v3/setup/etcdctl-auth.sh]: remove this if not being used

File: v3/setup/etcdctl-auth.sh [hunk 22]
```diff
@@ -0,0 +1,82 @@
...
+
+#Enable authentication
+
+curl  -L http://127.0.0.1:2379/v2/auth/enable -XPUT
+
+#etcdctl auth enable
+
+#Get authentication Status
+#curl -u $(echo "$(cat root.json |jq '.user')" | sed -e 's/^"//'  -e 's/"$//'):$(echo "$(cat root.json |jq '.password')" | sed -e 's/^"//'  -e 's/"$//') -L http://127.0.0.1:2379/v2/auth/users -XGET
```
Comment (chr0n1x) [inline v3/setup/etcdctl-auth.sh]: also remove this

File: v3/setup/etcdctl-auth.sh [hunk 23]
```diff
@@ -0,0 +1,82 @@
...
+
+curl  -L http://127.0.0.1:2379/v2/auth/users/$(eval echo $WRITE_USERNAME) -XPUT -d "@$CRED_DIR/readwriterole.json"
+
+#revoke guest role read:write access
+
+#etcdctl role revoke guest -path '/*' -readwrite
+
+#Enable authentication
+
+curl  -L http://127.0.0.1:2379/v2/auth/enable -XPUT
```
Comment (chr0n1x) [inline v3/setup/etcdctl-auth.sh]: extra whitespace between `curl` & `-L`

File: v3/setup/etcdctl-auth.sh [hunk 24]
```diff
@@ -0,0 +1,82 @@
...
+sudo echo '{"user":"'$(eval echo $WRITE_USERNAME)'", "password":"'$(eval echo $WRITE_PASSWORD)'"}' > $CRED_DIR/write-user.json
+curl  -L http://127.0.0.1:2379/v2/auth/users/$(eval echo $WRITE_USERNAME) -XPUT -d "@$CRED_DIR/write-user.json"
+
+#Create a read only role:
+
+etcdctl role add readonlyrole
+
+#Give read only access to this role:
+
+etcdctl role grant readonlyrole -path '/*' -read
```
Comment (chr0n1x) [inline v3/setup/etcdctl-auth.sh]: organize these lines please. That is, something more like:

```
# read-only access
etcdctl role add read-only
etcdctl role grant read-only -path '/*' -read

# read-write
etcdctl role add read-write
etcdctl role grant read-write -path '/*' -readwrite

# ...etc
```

Too much whitespace just makes it hard to read.

**Comment by chr0n1x** [comment]
please squash all of these commits into a single commit

**Comment by anshupande** [comment]
Moving this from my fork: https://github.com/adobe-platform/mesos-systemd/pull/206

**Pull request closed without merging**